### PR TITLE
Do not allow `reshape` on empty arrays

### LINKED ISSUE #2637: A size-0 array should not be reshaped into a size-1 array
It seems the `reshape()` function missed a size check somewhere?
```python
>>> import cupy as cp
>>> a = cp.zeros((0,))
>>> a.size
0
>>> a = a.reshape(())
>>> a.size
1
>>> a.shape
()
```
NumPy behaves as expected:
```python
>>> import numpy as np
>>> a = np.zeros((0,))
>>> a.size
0
>>> a = a.reshape(())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot reshape array of size 0 into shape ()
```

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #2637

### CODE REVIEW

File: cupy/core/_routines_manipulation.pyx [hunk 1]
```diff
@@ -130,6 +130,10 @@ cdef ndarray _ndarray_reshape(ndarray self, tuple shape, order):
     if len(shape) == 1 and cpython.PySequence_Check(shape[0]):
         shape = tuple(shape[0])
 
+    if self.size == 0:
```
Comment (kmaehashi) [inline cupy/core/_routines_manipulation.pyx]: `xp.array([]).reshape((0,))` should success.

Comment (emcastillo) [reply cupy/core/_routines_manipulation.pyx]: Good catch!

**Comment by emcastillo** [comment]
PTAL

File: cupy/core/_routines_manipulation.pyx [hunk 2]
```diff
@@ -333,6 +333,13 @@ cpdef ndarray _reshape(ndarray self,
     if internal.vector_equal(shape, self._shape):
         return self.view()
 
+    cdef size_t shape_size = 1
+    for d in shape:
+        shape_size *= d
```
Comment (kmaehashi) [inline cupy/core/_routines_manipulation.pyx]: You can use `internal.prod` here.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Thanks, @emcastillo!

**Pull request merged**